Refresh and simplify authentication/authorization samples (#331)
SteeltoeOSS/Samples: Security/src/AuthClient/README.md

# Steeltoe Application Security Client-side Authentication and Authorization

This application shows how to use the Steeltoe [security libraries](https://docs.steeltoe.io/api/v3/security/) for authentication and authorization with OpenID Connect against [Single Sign-On for VMware Tanzu Application Service](https://docs.vmware.com/en/Single-Sign-On-for-VMware-Tanzu-Application-Service) and using client certificates provided by Cloud Foundry or Steeltoe (when running locally).

## General pre-requisites

1. Installed .NET 8 SDK
1. Optional: [VMware Tanzu Platform for Cloud Foundry](https://docs.vmware.com/en/VMware-Tanzu-Application-Service/index.html)
   (optionally with [Windows support](https://docs.vmware.com/en/VMware-Tanzu-Application-Service/6.0/tas-for-vms/concepts-overview.html))
   with [Single Sign-On for VMware Tanzu Application Service](https://docs.vmware.com/en/Single-Sign-On-for-VMware-Tanzu-Application-Service)
   and [Cloud Foundry CLI](https://docs.cloudfoundry.org/cf-cli/install-go-cli.html)

## Running locally

1. Start a UAA Server [docker container](https://github.com/SteeltoeOSS/Samples/blob/main/CommonTasks.md)

> 1. `dotnet run`

Author: I'll actually just remove the http profiles entirely. They were useful during dev time, but I don't think they're needed anymore

> 1. Don't forget to start the [AuthServer](../AuthServer/README.md) application

Author: revised

> Note: To see the logs on Cloud Foundry as the app runs, execute this command: `cf logs steeltoe-samples-authclient`

From the website's menu, click on the `Log in` menu item and you should be redirected to the UAA server's login page. Enter `testuser` and `password`, and you should be authenticated and redirected back to the auth client home page.

The menu of the application includes links for testing the permissions of the user in the current application and interact with another service that has been secured with JWT and client certificates.

* The JWT menu item uses the current user's token to communicate with the backing service.
* The "SameSpace" and "SameOrg" menu items interact with the backing service using an identity certificate that belongs to the application.
   * Locally, certificates for both the client and server are created by Steeltoe.
   * On Cloud Foundry, certificates are provisioned by the platform, with OrgId and SpaceId populated based on where the applications are deployed.
* While logged in, view information about the testuser account by clicking on "Hello testuser!" next to the "Log off".

> 

Author: adding a note: `If needed, sign out of the UAA server using the dropdown menu in the top right corner at <http://localhost:8080>`

> 

Author: We also need some instructions for this scenario in a CF environment. There's a step missing that involves permission mapping and I apparently can't figure it out just by poking at the settings, so I'm going to have to go back to the docs and sort out exactly what to put where for the `sampleapi.read` to work when the sso tile is in the mix

1. Deploy [UAA with the Steeltoe Samples configuration](https://github.com/SteeltoeOSS/Dockerfiles/tree/main/uaa-server#customizing-for-your-environment)
1. Create a [service plan](https://docs.vmware.com/en/Single-Sign-On-for-VMware-Tanzu-Application-Service/1.14/sso/GUID-manage-service-plans.html)
1. Configure federated authentication on the service plan
   1. Add the UAA server from step 2 to the service as an [OIDC Provider](https://docs.vmware.com/en/Single-Sign-On-for-VMware-Tanzu-Application-Service/1.14/sso/GUID-configure-external-id.html#config-ext-oidc)
      * Name the identity provider `steeltoe-uaa` (or update `SSO_IDENTITY_PROVIDERS` in manifest.yml accordingly)
      * Credentials for connecting to the UAA server can be found or customized before deployment in [uaa.yml](https://github.com/SteeltoeOSS/Dockerfiles/blob/main/uaa-server/uaa.yml#L124)
   1. Save changes, but keep this page open
1. Create a service instance:
   * `cf create-service p-identity <your service plan name> mySSOService`
1. Push AuthServer to Cloud Foundry

>    1. `cf target -o myorg -s development`

Reviewer: ```suggestion
   1. `cf target -o your-org -s your-space`
```

   1. `cd samples/Security/src/AuthServer`
   1. `cf push`
     * When deploying to Windows, binaries must be built locally before push. Use the following commands instead:
     ```
     dotnet publish -r win-x64 --self-contained
     cf push -f manifest-windows.yml -p bin/Release/net8.0/win-x64/publish
     ```
1. Return to the service plan setup page and add an External Group Mapping with these values:
      * OIDC Groups Claim Name = scope
      * External Group Name = openid
      * Permissions = sampleapi.read (If this option isn't available, ensure AuthServer has been deployed)
1. Push AuthClient to Cloud Foundry

>    1. `cf target -o myorg -s development`

Reviewer: ```suggestion
   1. `cf target -o your-org -s your-space`
```

   1. `cd samples/Security/src/AuthClient`
   1. `cf push`
     * When deploying to Windows, binaries must be built locally before push. Use the following commands instead:
     ```
     dotnet publish -r win-x64 --self-contained
     cf push -f manifest-windows.yml -p bin/Release/net8.0/win-x64/publish
     ```

> > Note: The provided manifests will create apps named `steeltoe-samples-authclient` and `steeltoe-samples-authserver` and attempt to bind both to the SSO service `mySSOService`.

Reviewer: ```suggestion
> [!NOTE]
> The provided manifests will create apps named `steeltoe-samples-authclient` and `steeltoe-samples-authserver`
> and attempt to bind both to the SSO service `mySSOService`.
```

### RedirectUri and Scope access

The RedirectUri and Scope access settings should be automatically configured via the settings in `manifest.yml`.

If you want to access the `sso` dashboard, run the following command and go to the URL listed in `dashboard url` property:

```bash
$ cf service mySSOService

name:            mySSOService
guid:            ea8b8ac0-ce85-4726-8b39-d1b2eb55b45b
type:            managed
broker:          identity-service-broker
offering:        p-identity
plan:            steeltoe
tags:
offering tags:
description:     Provides identity capabilities via UAA as a Service
documentation:   https://docs.pivotal.io/p-identity/index.html
dashboard url:   https://p-identity.sys.cf-app.com/developer/identity-zones/15aaabfa-0697-4ad7-96a8-ed81c0a286a7/instances/ea8b8ac0-ce85-4726-8b39-d1b2eb55b45b/
...
```

## What to expect

At this point the app is up and running.  You can access it at <https://localhost:7072> or <https://steeltoe-samples-authclient.`YOUR-CLOUDFOUNDRY-APP-DOMAIN`/>.

> > Note: To see the logs on Cloud Foundry as the app runs, execute this command: `cf logs steeltoe-samples-authclient`

Reviewer: ```suggestion
> [!NOTE]
> To see the logs on Cloud Foundry as the app runs, execute this command: `cf logs steeltoe-samples-authclient`
```

From the website's menu, click on the `Log in` menu item and you should be redirected to the UAA server's login page. Enter `testuser` and `password`, and you should be authenticated and redirected back to the auth client home page.

The menu of the application includes links for testing the permissions of the user in the current application and interact with another service that has been secured with JWT and client certificates.

> * The JWT menu item uses the current user's token to communicate with the backing service.

Reviewer: ```suggestion
* The JWT menu item uses the current user's token to communicate with the backend service.
```

> * The "SameSpace" and "SameOrg" menu items interact with the backing service using an identity certificate that belongs to the application.

Reviewer: ```suggestion
* The "SameSpace" and "SameOrg" menu items interact with the backend service using an identity certificate that belongs to the application.
```

   * Locally, certificates for both the client and server are created by Steeltoe.
   * On Cloud Foundry, certificates are provisioned by the platform, with OrgId and SpaceId populated based on where the applications are deployed.

> * While logged in, view information about the testuser account by clicking on "Hello testuser!" next to the "Log off".

Reviewer: ```suggestion
* While logged in, view information about the testuser account by clicking on "Hello testuser!" next to the "Log out" link.
```
